src/Native/Unix/System.Native/pal_networking.cpp

```diff
@@ -2134,38 +2134,37 @@ SystemNative_SetSockOpt(intptr_t socket, int32_t socketOptionLevel, int32_t sock
     //
     if (socketOptionLevel == PAL_SOL_SOCKET)
     {
-        if (socketOptionName == PAL_SO_EXCLUSIVEADDRUSE)
+        // Windows supports 3 address sharing modes:
+        // - not sharing      (SO_EXCLUSIVEADDRUSE=1, SO_REUSEADDR=0)
+        // - explicit sharing (SO_EXCLUSIVEADDRUSE=0, SO_REUSEADDR=1)
+        // - implicit sharing (SO_EXCLUSIVEADDRUSE=0, SO_REUSEADDR=0)
+        // On Unix we have two address sharing modes:
+        // - not sharing      (SO_REUSEPORT=0)
+        // - explicit sharing (SO_REUSEPORT=1)
+        // We make both PAL_SO_REUSEADDR and PAL_SO_EXCLUSIVEADDRUSE control SO_REUSEPORT.
+        if (socketOptionName == PAL_SO_EXCLUSIVEADDRUSE || socketOptionName == PAL_SO_REUSEADDR)
         {
-            //
-            // SO_EXCLUSIVEADDRUSE makes Windows behave like Unix platforms do WRT the SO_REUSEADDR option.
-            // So, on Unix platforms, we consider SO_EXCLUSIVEADDRUSE to always be set.  We allow manually setting this
-            // to "true", but not "false."
-            //
             if (optionLen != sizeof(int32_t))
             {
                 return PAL_EINVAL;
             }
 
-            if (*reinterpret_cast<int32_t*>(optionValue) == 0)
-            {
-                return PAL_ENOTSUP;
-            }
-            else
-            {
-                return PAL_SUCCESS;
-            }
-        }
-        else if (socketOptionName == PAL_SO_REUSEADDR)
-        {
-            //
-            // On Windows, SO_REUSEADDR allows the address *and* port to be reused.  It's equivalent to 
-            // SO_REUSEADDR + SO_REUSEPORT other systems. 
-            //
-            int err = setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, optionValue, static_cast<socklen_t>(optionLen));
-            if (err == 0)
+            int value = *reinterpret_cast<int32_t*>(optionValue);
+
+            // PAL_SO_EXCLUSIVEADDRUSE is inverse of PAL_SO_REUSEADDR.
+            if (socketOptionName == PAL_SO_EXCLUSIVEADDRUSE)
             {
-                err = setsockopt(fd, SOL_SOCKET, SO_REUSEPORT, optionValue, static_cast<socklen_t>(optionLen));
+                if ((value != 0) && (value != 1))
```

Comment:
How does ReuseAddress behave when passing in a value other than 0 or 1?  Does it fail, or does it treat anything non-zero as being true?  I'm wondering if we want/need this if block at all.

Reply:
While doing this PR I learned mac returns 512 for getsockopt SO_REUSEPORT when doing setsockopt to 1 (https://github.com/dotnet/corefx/pull/24809/commits/cdac9c68138f9474afe753a86145083deb2eb893).
So behavior for getsockopt and setsockopt is platform dependent. Perhaps we should iron out some of those differences for other options too. I don't think we are limiting by restricting to 0 and 1.

Reply:
Ok.